### PR TITLE
Put build number in pre-release version instead of build metadata

### PR DESCRIPTION
**Pull request by martinholters**
Ref #17330.

If accepted, this PR should be merged soonish after VERSION is set to `0.6.0-dev`. (That's the assumption made in contrib/commit-name.sh.) Also, bin/version.sh in Compat would need adjustment. Surprisingly, I haven't found other places that would need to be changed. Surely I missed some?

Apart from switching from `0.6.0-dev+123` to `0.6.0-dev.123`, this PR handles the situation where the "build number" (=number of commits since latest tag) is zero (which usually means could not be determined) differently. Instead of omitting it, it adds `.unknown` if there is a pre-release version part in VERSION. E.g. `0.6.0` stays `0.6.0`, but `0.6.0-dev` becomes `0.6.0-dev.unknown`. OTOH, if VERSION is not a pre-release version but the "build number" is non-zero, it is ignored (and a corresponding message is printed). Would this be the right thing to do during preparation of a release?

CC @tkelman

### CODE REVIEW

**Comment by tkelman** [comment]
>  OTOH, if VERSION is not a pre-release version but the "build number" is non-zero

I think the only way that would happen would be if some commits were made after dropping -pre from VERSION, but before adding it back for the next release. This shouldn't happen if we're careful about the release branch.

I guess we'd have to be on the lookout for people still using +123 after this gets merged, and fix any occurrences of that for 0.6.0-dev or later.

How does dev.unknown compare to known pre.123?

**Comment by tkelman** [comment]
also double check version_git.sh just to be sure. not sure how to test this.

File: contrib/commit-name.sh [hunk 1]
```diff
@@ -8,5 +8,11 @@ gitref=${1:-HEAD}
...
 last_tag=$(git describe --tags --abbrev=0 "$gitref")
 ver=$(git show "$gitref:VERSION")
+major=$(echo $ver | cut -f 1 -d .)
+minor=$(echo $ver | cut -f 2 -d .)
 nb=$(git rev-list --count "$gitref" "^$last_tag")
-echo "$ver+$nb"
+if [ $major = 0 -a $minor -le 5 ]; then
+  echo "$ver+$nb"
+else
+  echo "$ver.$nb"
```
Comment (tkelman) [inline contrib/commit-name.sh]: I guess both of these (including the current script) are wrong if you happen to run it on a tag?

**Comment by martinholters** [comment]
> > OTOH, if VERSION is not a pre-release version but the "build number" is non-zero

>I think the only way that would happen would be if some commits were made after dropping -pre from VERSION, but before adding it back for the next release. This shouldn't happen if we're careful about the release branch.

I could imagine this scenario: Drop `-pre` from VERSION, maybe commit or maybe not, but build before tagging - maybe just locally for testing.

>I guess we'd have to be on the lookout for people still using +123 after this gets merged, and fix any occurrences of that for 0.6.0-dev or later.

Yes, that would be the obvious downside of accepting this PR.

>How does dev.unknown compare to known pre.123

`v"0.6.0-dev" < v"0.6.0-dev.123" < v"0.6.0-dev.unknown" < v"0.6.0-pre" < v"0.6.0-pre.123" < v"0.6.0-pre.unknown"`

>also double check version_git.sh just to be sure. not sure how to test this.

IIUC it only gathers the raw data from git but does not have direct contact with the version numbers.

File: contrib/commit-name.sh (hunk 1 above)
Comment (martinholters) [inline contrib/commit-name.sh]: Ah yes, should be just `echo $ver` if `$nb` is zero, I guess. Good opportunity to fix; will do next week.

Maybe we could eventually rewrite these scripts in a nicer programming language, like, say, Julia.

**Comment by tkelman** [comment]
> `v"0.6.0-dev" < v"0.6.0-dev.123" < v"0.6.0-dev.unknown" < v"0.6.0-pre" < v"0.6.0-pre.123" < v"0.6.0-pre.unknown"`

can we throw in some tests to verify this kind of thing, if the existing version number comparisons don't already cover these?

File: contrib/commit-name.sh (hunk 1 above)
Comment (martinholters) [inline contrib/commit-name.sh]: > Good opportunity to fix; will do next week.

Done.

**Comment by martinholters** [comment]
>can we throw in some tests to verify this kind of thing

Done.

**Comment by martinholters** [comment]
Orthogonal to this PR as it is now, but related: Should the "build number" maybe be based on the commit VERSION was last changed instead of on the latest tag that can be found? Seems a bit more robust to me. (If I tag some commit locally as kind of a bookmark or whatever, that should not screw up the version in my builds, should it?)

**Comment by tkelman** [comment]
Looking at changes to VERSION might indeed be more robust than looking at "most recent tag." We have had instances before of people making local tags and that screwing things up.

**Comment by martinholters** [comment]
Also in light of https://github.com/JuliaLang/julia/issues/17418#issuecomment-233871337, I propose to open a separte PR that uses `git blame VERSION` or the like instead of `git describe --tags` and includes the two probably uncontroversial commits from this PR.

**Comment by martinholters** [comment]
Rebased on top of #17531, so this should be ready for merge after branching ... if we want it, that is.

**Comment by tkelman** [comment]
We'll need the equivalent of https://github.com/JuliaLang/Compat.jl/issues/259, right? Is `commit-name.sh` a straight copy-paste?

**Pull request merged**